[Maintenance][Statistics] Bit of an overhaul

NoResponseMate opened the pull request:
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13 |
| Bug fix?        | no                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no |
| Related tickets | -                      |
| License         | MIT                                                          |

NoResponseMate committed:
[Maintenance][Statistics] Simplify get action

NoResponseMate committed:
[Maintenance][Statistics] Remove overreaching validation

NoResponseMate committed:
[Maintenance][Statistics] Simplify input validation

NoResponseMate committed:
[Maintenance] Refactor retrieving orders statistics logic to be more customizable

NoResponseMate committed:
[Maintenance] Simplify sales statistics and make the return structure more agnostic

NoResponseMate committed:
[Maintenance] Ignore abstract classes in spec arkitect rule

NoResponseMate committed:
[API][Admin] A bit more statistics tests

NoResponseMate committed:
[Maintenance][Statistics] Force same type while comparing totals' years

jakubtobiasz commented on src/Sylius/Component/Core/spec/Statistics/Provider/OrdersTotals/YearBasedOrdersTotalProviderSpec.php:
```diff
@@ -0,0 +1,102 @@
+        ]);
+    }
+
+    function it_returns_fills_zeros_in_periods_with_no_totals(
```
Something went wrong here I guess :D

jakubtobiasz commented on src/Sylius/Component/Core/spec/Statistics/Provider/OrdersTotals/DayBasedOrdersTotalProviderSpec.php:
```diff
@@ -0,0 +1,104 @@
+        $this->shouldImplement(OrdersTotalsProviderInterface::class);
+    }
+
+    function it_returns_an_array_of_zeros_when_no_totals_have_been_found_for_period(
```
Maybe something like `it returns zero as total when no totals have been found` or something like that. There is a single zero.

NoResponseMate replied:
Added more zeros 💃

GSadee commented on src/Sylius/Bundle/ApiBundle/Controller/GetStatisticsAction.php:
```diff
@@ -23,29 +22,29 @@
 use Symfony\Component\Messenger\MessageBusInterface;
 use Symfony\Component\Serializer\SerializerInterface;
-use Symfony\Component\Validator\Constraints as BaseAssert;
+use Symfony\Component\Validator\Constraints as SymfonyAssert;
```
```suggestion
use Symfony\Component\Validator\Constraints as SymfonyConstraints;
```

GSadee commented on src/Sylius/Bundle/ApiBundle/Controller/GetStatisticsAction.php:
```diff
@@ -13,7 +13,6 @@
 
-use Sylius\Bundle\ApiBundle\Exception\ChannelNotFoundException;
 use Sylius\Bundle\ApiBundle\Query\GetStatistics;
 use Sylius\Bundle\ApiBundle\Validator\Constraints as Assert;
```
```suggestion
use Sylius\Bundle\ApiBundle\Validator\Constraints;
```

GSadee commented on src/Sylius/Bundle/ApiBundle/OpenApi/Documentation/StatisticsDocumentationModifier.php:
```diff
@@ -37,42 +39,46 @@ public function modify(OpenApi $docs): OpenApi
                             ],
                             'total' => [
                                 'type' => 'integer',
                                 'example' => '1000',
```
```suggestion
                                'example' => 1000,
```
as it is an integer?

GSadee commented on src/Sylius/Bundle/CoreBundle/Attribute/AsOrdersTotalsProvider.php:
```diff
@@ -0,0 +1,33 @@
+#[\Attribute(\Attribute::TARGET_CLASS)]
+final class AsOrdersTotalsProvider
+{
+    public const SERVICE_TAG = 'sylius.statistics.orders_totals_provider';
```
There is some inconsistency when it comes to tags, because in the service definitions, you used: `sylius.orders_totals_provider` without `statistics` part.

NoResponseMate replied:
I've changed the config tags

GSadee commented on src/Sylius/Component/Core/spec/Statistics/Provider/SalesStatisticsProviderSpec.php:
```diff
@@ -0,0 +1,119 @@
+        $this->shouldImplement(SalesStatisticsProviderInterface::class);
+    }
+
+    function it_throw_exception_when_period_format_for_interval_type_is_unknown(
```
```suggestion
    function it_throws_an_exception_when_period_format_for_interval_type_is_unknown(
```

NoResponseMate committed:
Review fixes
[Maintenance] Make order totals providers specs less dependant on current datetime

Rafikooo commented on src/Sylius/Bundle/CoreBundle/Resources/config/services/providers.xml:
```diff
@@ -37,18 +37,33 @@
         </service>
 
-        <service id="Sylius\Component\Core\Statistics\Provider\SalesTimeSeriesProviderInterface" class="Sylius\Bundle\CoreBundle\Provider\SalesTimeSeriesProvider">
+        <service id="Sylius\Component\Core\Statistics\Provider\SalesSummaryProviderInterface" class="Sylius\Component\Core\Statistics\Provider\BusinessActivitySummaryProvider">
```
I believe this service ID should correspond to their class
```suggestion
        <service id="Sylius\Component\Core\Statistics\Provider\BusinessActivitySummaryProviderInterface" class="Sylius\Component\Core\Statistics\Provider\BusinessActivitySummaryProvider">
```

jakubtobiasz commented on src/Sylius/Bundle/ApiBundle/Controller/GetStatisticsAction.php:
```diff
@@ -96,4 +92,36 @@ private function createBadRequestResponse(ConstraintViolationListInterface $viol
         );
     }
+
+    private function createInputDataConstraints(): SymfonyConstraints\Collection
+    {
+        return new SymfonyConstraints\Collection([
+            'channelCode' => new Constraints\Code(),
+            'startDate' => [
+                new SymfonyConstraints\NotBlank(),
+                new SymfonyConstraints\DateTime('Y-m-d\TH:i:s', message: 'sylius.date_time.invalid'),
+            ],
+            'interval' => new SymfonyConstraints\Choice(choices: array_keys($this->intervalsMap), multiple: false),
+            'endDate' => [
+                new SymfonyConstraints\NotBlank(),
+                new SymfonyConstraints\DateTime('Y-m-d\TH:i:s', message: 'sylius.date_time.invalid'),
+                new SymfonyConstraints\GreaterThan(propertyPath: 'startDate'),
+            ],
+        ]);
+    }
```
I have mixed feelings about this. It looks like a hard thing to be modified.

jakubtobiasz commented:
Thank you, @NoResponseMate!
